[DTensor] migrate prop_index (sharding prop rule for op aten.index.Tensor) to sharding prop strategy

XilunWu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #159003

**Summary**
migrate aten.index.Tensor from prop rule to prop strategy. This PR first supports
acceptance of full-replicate shardings on `self` and `indices`.

**Test**
`pytest test/distributed/tensor/test_tensor_ops.py -s -k index`

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @msaroufim @dcci @tianyu-l @SherlockNoMad

XilunWu committed:
[1/N][DTensor] migrate prop_index: support all-replicate on aten.index.Tensor

[ghstack-poisoned]

wanchaol reviewed (changes requested):
hmmm migrating to an all-replicate strategy would break the existing behavior.. existing rules support sharding, so migration should support the same feature

XilunWu commented on torch/distributed/tensor/_op_schema.py:
```diff
@@ -351,7 +351,12 @@ def __str__(self) -> str:
                 mesh_shape = arg.mesh_shape
             elif isinstance(arg, TupleStrategy):
-                first_op_strategy = arg.children[0]
+                first_op_strategy_idx = 0
```
There's no clear contract on that, and I prefer that `TupleStrategy` shouldn't. However, we don't have a good sharding structure for `Sequence[Optional[Tensor]]` argument.

wconstab commented on torch/distributed/tensor/_ops/_tensor_ops.py:
I think one PR is better. If you need to change behavior of tuplestrategy that could be a separate Pr with a test, but see my other comment on that matter.

ezyang commented:
Isn't porting a prop rule to a prop strategy purely mechanical? You just iterate over the OpStrategy strategies to get the OpSchema, run the prop rule as before, and then collect all the OutputShardings up into the output strategy? In fact, shouldn't there be a generic function we can write that transforms a prop rule into a prop strategy?

XilunWu committed:
Update on "[1/N][DTensor] migrate prop_index: support all-replicate on aten.index.Tensor"


**Summary**
migrate aten.index.Tensor from prop rule to prop strategy. This PR first supports
acceptance of full-replicate shardings on `self` and `indices`.

**Test**
`pytest test/distributed/tensor/test_tensor_ops.py -s -k index`

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta tianyu-l

[ghstack-poisoned]

XilunWu committed:
Update on "[1/N][DTensor] migrate prop_index: support all-replicate on aten.index.Tensor"


**Summary**
migrate aten.index.Tensor from prop rule to prop strategy. This PR first supports
acceptance of full-replicate shardings on `self` and `indices`.

**Test**
`pytest test/distributed/tensor/test_tensor_ops.py -s -k index`

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta tianyu-l

[ghstack-poisoned]

XilunWu committed:
Update on "[DTensor] migrate prop_index: support all-replicate on aten.index.Tensor"


**Summary**
migrate aten.index.Tensor from prop rule to prop strategy. This PR first supports
acceptance of full-replicate shardings on `self` and `indices`.

**Test**
`pytest test/distributed/tensor/test_tensor_ops.py -s -k index`

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta tianyu-l

[ghstack-poisoned]

XilunWu committed:
Update on "[DTensor] migrate prop_index: support all-replicate on aten.index.Tensor"


**Summary**
migrate aten.index.Tensor from prop rule to prop strategy. This PR first supports
acceptance of full-replicate shardings on `self` and `indices`.

**Test**
`pytest test/distributed/tensor/test_tensor_ops.py -s -k index`

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta tianyu-l

[ghstack-poisoned]

XilunWu committed:
Update on "[DTensor] migrate prop_index: support all-replicate on aten.index.Tensor"


**Summary**
migrate aten.index.Tensor from prop rule to prop strategy. This PR first supports
acceptance of full-replicate shardings on `self` and `indices`.

**Test**
`pytest test/distributed/tensor/test_tensor_ops.py -s -k index`

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta tianyu-l

[ghstack-poisoned]

XilunWu committed:
Update on "[DTensor] migrate prop_index: support all-replicate on aten.index.Tensor"


**Summary**
migrate aten.index.Tensor from prop rule to prop strategy. This PR first supports
acceptance of full-replicate shardings on `self` and `indices`.

**Test**
`pytest test/distributed/tensor/test_tensor_ops.py -s -k index`

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta tianyu-l

[ghstack-poisoned]

XilunWu committed:
Update on "[DTensor] migrate prop_index: support all-replicate on aten.index.Tensor"


**Summary**
migrate aten.index.Tensor from prop rule to prop strategy. This PR first supports
acceptance of full-replicate shardings on `self` and `indices`.

**Test**
`pytest test/distributed/tensor/test_tensor_ops.py -s -k index`

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta tianyu-l

[ghstack-poisoned]

XilunWu committed:
Update on "[DTensor] migrate prop_index: support all-replicate on aten.index.Tensor"


**Summary**
migrate aten.index.Tensor from prop rule to prop strategy. This PR first supports
acceptance of full-replicate shardings on `self` and `indices`.

**Test**
`pytest test/distributed/tensor/test_tensor_ops.py -s -k index`

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta tianyu-l

[ghstack-poisoned]
